### DISCUSSION TITLE
furo 2022.6.4.1: AttributeError: 'document' object has no attribute 'findall'

### DISCUSSION THREAD

**Discussion by sunpoet** [pradyunsg/furo#453] [Potential Issue] [answered]
While building pip document with furo 2022.6.4.1, I get:
```
  File "/usr/local/lib/python3.8/site-packages/furo/__init__.py", line 47, in run
    for node in list(self.document.findall(nodes.table)):
AttributeError: 'document' object has no attribute 'findall'
```
It was add in 7fa8d083a1f6630e06b629f2cfc319a2a5b806d1.

And the build process works fine with furo 2022.4.7.

**Comment by pradyunsg** [comment]
What version of Sphinx are you on?

**Reply by sunpoet** [reply]
4.3.1

**Comment by chrisjsewell** [comment, marked as answer]
`findall` is only added in docutils 0.18. In myst-parser I added a compatibility function, to fallback to `traverse` on an attribute error; https://github.com/executablebooks/MyST-Parser/blob/75ef9cb7b65c98d969724fc6c096e8d6209c5ea0/myst_parser/_compat.py#L7

**Reply by pradyunsg** [reply]
Ah, fun.

**Reply by sunpoet** [reply]
No wonder it fails on my side. We're still using [docutils 0.17.1](https://www.freshports.org/textproc/py-docutils).

**Comment by rbetz** [comment]
I also just encountered this error.

```python
File "/Users/robin/miniconda/lib/python3.9/site-packages/furo/__init__.py", line 47, in run
  for node in list(self.document.findall(nodes.table)):
AttributeError: 'document' object has no attribute 'findall'
```

I see that the `findall` feature was added in `docutils==0.18.1`.
However, Sphinx 4 requires `docutils>=0.14, < 0.18`. and 0.18 support was added in Sphinx 5 [(changelog here)](https://www.sphinx-doc.org/en/master/changes.html). Indeed, my machine ends up with `docutils==0.17.1` installed.

Possible solutions:
 - Require Sphinx 5 or greater
 - Minor code refactor to use `Node.traverse` instead of `Node.findall`

**Reply by pradyunsg** [reply]
Can you try doing `pip install https://github.com/pradyunsg/furo/archive/refs/heads/main.zip`, and confirming that it works?

**Reply by rbetz** [reply]
That works!

**Reply by sunpoet** [reply]
2022.6.21 works for me. BTW, there's a typo in docs/changelog.md where date is "2022.07.21".

**Reply by pradyunsg** [reply]
Whelp, I can't count months.
